### PR TITLE
Change Ska.Sun to ska_sun w/ back-compatibility

### PR DESCRIPTION
**Pull request by taldcroft**
## Description

This is a proof of concept to change the `Ska.Sun` package to natively be named `ska_sun`. The key here is adjusting three `setup()` inputs so that setuptools makes a complete duplicate of `ska_sun` within the `Ska.Sun` namespace for full back-compatibility. This seems to work reliably, the downside being duplication of files and data in the `ska_sun` distribution.

I played around with some ideas to make a lightweight virtual copy using dynamic import management / hacking, but I could not make this work.

Although not strictly necessary, the hope would be to change the repo name to `ska_sun` as well.

This strategy would then be applied to all the `Ska.*` and `Chandra.*` repos.

## Interface impacts

Adds a new alias for importing packages. Changing the repo name should not impact users but will require corresponding changes in `skare3`.

### Unit tests

- [x] Mac

Independent check of unit tests by Jean (I did a --user install instead of a within-environment install)
- [x] Linux

#### Build and install
```
(ska3-dev) ➜  Ska.Sun git:(de-namespace) python setup.py install --single-version-externally-managed --record=record.txt
running install
running build
running build_py
creating build
creating build/lib
creating build/lib/ska_sun
copying ska_sun/sun.py -> build/lib/ska_sun
copying ska_sun/__init__.py -> build/lib/ska_sun
creating build/lib/ska_sun/tests
copying ska_sun/tests/__init__.py -> build/lib/ska_sun/tests
copying ska_sun/tests/test_sun.py -> build/lib/ska_sun/tests
creating build/lib/Ska
creating build/lib/Ska/Sun
copying ska_sun/sun.py -> build/lib/Ska/Sun
copying ska_sun/__init__.py -> build/lib/Ska/Sun
creating build/lib/Ska/Sun/tests
copying ska_sun/tests/__init__.py -> build/lib/Ska/Sun/tests
copying ska_sun/tests/test_sun.py -> build/lib/Ska/Sun/tests
creating build/lib/ska_sun/data
copying ska_sun/data/pitch_roll.fits.gz -> build/lib/ska_sun/data
creating build/lib/Ska/Sun/data
copying ska_sun/data/pitch_roll.fits.gz -> build/lib/Ska/Sun/data
running install_lib
copying build/lib/Ska/Sun/sun.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun
copying build/lib/Ska/Sun/tests/__init__.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests
copying build/lib/Ska/Sun/tests/test_sun.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests
copying build/lib/Ska/Sun/__init__.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun
copying build/lib/Ska/Sun/data/pitch_roll.fits.gz -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/data
copying build/lib/ska_sun/sun.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun
copying build/lib/ska_sun/tests/__init__.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests
copying build/lib/ska_sun/tests/test_sun.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests
copying build/lib/ska_sun/__init__.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun
copying build/lib/ska_sun/data/pitch_roll.fits.gz -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/data
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/sun.py to sun.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests/__init__.py to __init__.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests/test_sun.py to test_sun.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/__init__.py to __init__.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/sun.py to sun.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/__init__.py to __init__.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/test_sun.py to test_sun.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/__init__.py to __init__.cpython-38.pyc
running install_egg_info
running egg_info
creating ska_sun.egg-info
writing ska_sun.egg-info/PKG-INFO
writing dependency_links to ska_sun.egg-info/dependency_links.txt
writing top-level names to ska_sun.egg-info/top_level.txt
writing manifest file 'ska_sun.egg-info/SOURCES.txt'
adding license file 'LICENSE.rst'
writing manifest file 'ska_sun.egg-info/SOURCES.txt'
Copying ska_sun.egg-info to /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun-3.8.2.dev3+g75f1b72-py3.8.egg-info
running install_scripts
writing list of installed files to 'record.txt'
```
#### Test installed version
```
(ska3-dev) ➜  Ska.Sun git:(de-namespace) cd docs 

(ska3-dev) ➜  docs git:(de-namespace) ipython
Python 3.8.12 (default, Oct 12 2021, 06:23:56) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.29.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import Ska.Sun

In [2]: Ska.Sun.__file__
Out[2]: '/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/__init__.py'

In [3]: Ska.Sun.__version__
Out[3]: '3.8.2.dev3+g75f1b72'

In [4]: Ska.Sun.test()
============================================================================ test session starts ============================================================================
platform darwin -- Python 3.8.12, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages
plugins: remotedata-0.3.3, doctestplus-0.11.2, arraydiff-0.3, anyio-2.2.0, mock-3.6.1, filter-subpackage-0.1.1, openfiles-0.5.0, astropy-header-0.1.2, cov-3.0.0
collected 8 items                                                                                                                                                           

Ska/Sun/tests/test_sun.py ........                                                                                                                                    [100%]

============================================================================= 8 passed in 0.06s =============================================================================
Out[4]: False

In [5]: import ska_sun

In [6]: ska_sun.__file__
Out[6]: '/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/__init__.py'

In [7]: ska_sun.__version__
Out[7]: '3.8.2.dev3+g75f1b72'

In [8]: ska_sun.test()
============================================================================ test session starts ============================================================================
platform darwin -- Python 3.8.12, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages
plugins: remotedata-0.3.3, doctestplus-0.11.2, arraydiff-0.3, anyio-2.2.0, mock-3.6.1, filter-subpackage-0.1.1, openfiles-0.5.0, astropy-header-0.1.2, cov-3.0.0
collected 8 items                                                                                                                                                           

ska_sun/tests/test_sun.py ........                                                                                                                                    [100%]

============================================================================= 8 passed in 0.05s =============================================================================
Out[8]: False
```

### CODE REVIEW

**Review by jzuhone — APPROVED** [review]

**Comment by jeanconn** [comment]
I don't know if this will be an issue for us, but I note that while Ska.Sun.__version__ is defined, Ska.Sun.__file__ is not.

**Comment by taldcroft** [comment]
> I don't know if this will be an issue for us, but I note that while Ska.Sun.__version__ is defined, Ska.Sun.__file__ is not.

That doesn't match what I'm seeing. How did you test that? In particular, `/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/__init__.py` does get installed.
```
(ska3-dev) ➜  Ska.Sun git:(de-namespace) python setup.py install --single-version-externally-managed --record=record.txt                   
running install
running build
running build_py
copying ska_sun/sun.py -> build/lib/ska_sun
copying ska_sun/__init__.py -> build/lib/ska_sun
copying ska_sun/sun.py -> build/lib/Ska/Sun
copying ska_sun/__init__.py -> build/lib/Ska/Sun
running install_lib
copying build/lib/Ska/Sun/sun.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun
copying build/lib/Ska/Sun/__init__.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun
creating /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun
copying build/lib/ska_sun/sun.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun
creating /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests
copying build/lib/ska_sun/tests/__init__.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests
copying build/lib/ska_sun/tests/test_sun.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests
copying build/lib/ska_sun/__init__.py -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun
creating /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/data
copying build/lib/ska_sun/data/pitch_roll.fits.gz -> /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/data
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/sun.py to sun.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/__init__.py to __init__.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/test_sun.py to test_sun.cpython-38.pyc
byte-compiling /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/__init__.py to __init__.cpython-38.pyc
running install_egg_info
running egg_info
writing ska_sun.egg-info/PKG-INFO
writing dependency_links to ska_sun.egg-info/dependency_links.txt
writing top-level names to ska_sun.egg-info/top_level.txt
adding license file 'LICENSE.rst'
writing manifest file 'ska_sun.egg-info/SOURCES.txt'
Copying ska_sun.egg-info to /Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun-3.8.2.dev2+g05fe28e-py3.8.egg-info
running install_scripts
writing list of installed files to 'record.txt'
(ska3-dev) ➜  Ska.Sun git:(de-namespace) cat record.txt 
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/sun.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/__init__.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/__init__.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/test_sun.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/sun.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/__init__.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests/__init__.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests/test_sun.py
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/data/pitch_roll.fits.gz
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/data/pitch_roll.fits.gz
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/__pycache__/sun.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/__pycache__/__init__.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/__pycache__/__init__.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/tests/__pycache__/test_sun.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/__pycache__/sun.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/__pycache__/__init__.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests/__pycache__/__init__.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/tests/__pycache__/test_sun.cpython-38.pyc
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun-3.8.2.dev2+g05fe28e-py3.8.egg-info/PKG-INFO
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun-3.8.2.dev2+g05fe28e-py3.8.egg-info/not-zip-safe
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun-3.8.2.dev2+g05fe28e-py3.8.egg-info/SOURCES.txt
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun-3.8.2.dev2+g05fe28e-py3.8.egg-info/top_level.txt
/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun-3.8.2.dev2+g05fe28e-py3.8.egg-info/dependency_links.txt
(ska3-dev) ➜  Ska.Sun git:(de-namespace) cd
(ska3-dev) ➜  ~ ipython
Python 3.8.12 (default, Oct 12 2021, 06:23:56) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.29.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ska_sun

In [2]: import Ska.Sun

In [3]: Ska.Sun.__file__
Out[3]: '/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/Ska/Sun/__init__.py'

In [4]: ska_sun.__file__
Out[4]: '/Users/aldcroft/miniconda3/envs/ska3-dev/lib/python3.8/site-packages/ska_sun/__init__.py'
```

**Comment by jeanconn** [comment]
I'd done this
```
ska3-jeanconn-fido> pip install --user .
Processing /proj/sot/ska/jeanproj/git/Ska.Sun
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
Building wheels for collected packages: ska-sun
  Building wheel for ska-sun (setup.py) ... done
  Created wheel for ska-sun: filename=ska_sun-3.8.2.dev2+g05fe28e.d20221217-py3-none-any.whl size=120988 sha256=7f89e6bf866e957293799366ae535e9642626c7071000844bd106ea5b6b57ecb
  Stored in directory: /tmp/pip-ephem-wheel-cache-u_iv0n06/wheels/f7/19/98/a310debe4f30dcf198399a33fc13dd2e63ae6b388bf04355d4
Successfully built ska-sun
Installing collected packages: ska-sun
Successfully installed ska-sun-3.8.2.dev2+g05fe28e.d20221217
ska3-jeanconn-fido> cd
ska3-jeanconn-fido> ipython
Python 3.8.12 (default, Oct 12 2021, 13:49:34) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.29.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import Ska.Sun

In [2]: import Ska.Sun.__file__
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-2-5775e1be7a26> in <module>
----> 1 import Ska.Sun.__file__

ModuleNotFoundError: No module named 'Ska.Sun.__file__'
```

**Comment by jeanconn** [comment]
Works now so I'm not sure what happened.  I assumed it generalized but I guess I was mistaken.

**Comment by taldcroft** [comment]
`pip install` is a hard fail with namespace packages in my experience. Just don't do that.

**Comment by jeanconn** [comment]
I'd just copied from the functional tests above.

**Comment by taldcroft** [comment]
Oops. The hard fail is that `pip` removes every other package in that namespace. I guess for a `--user` install that normally doesn't matter, except maybe you discovered some other problem.

**Comment by jeanconn** [comment]
Yeah, I didn't discover much just because I did the same thing (as far as I can tell) and now have a different result.  So I'll assume something weird with "pip" in this case.

**Comment by javierggt** [comment]
I tried both with our standard:
```
python setup.py install --single-version-externally-managed --record=record.txt
```
and a typical way I like to test things:
```
pip install -e .
```

The first one is ok. In the second one the namespace package is not installed. Not an issue, but good to know.

I went over the skare3 building, and I think the only thing we will have to do is to add `testr` and `ska_helpers` to the recipes. While checking that, I noticed that many packages have the following in setup.py:
```
try:
    from testr.setup_helper import cmdclass
except ImportError:
    cmdclass = {}
```
but not a single recipe includes testr as a build dependency, so this is never used in the actual builds.

**Comment by jeanconn** [comment]
But we also don't run the tests when we build, so does it matter?

**Comment by javierggt** [comment]
> But we also don't run the tests when we build, so does it matter?

This PR adds this line:
```
from testr.setup_helper import cmdclass
```
and this will fail in the conda build unless testr is a build dependency. It could add it like the other packages (what I pasted above) and then testr will not be required as a build dependency. In either case, ska_helpers will need to be added as a build dependency in all recipes.

**Comment by taldcroft** [comment]
I wasn't sure about making that change to always require `testr`. I mostly got fed up with it looking ugly. :smile:. Is it better if I revert that or bite the bullet and change the build recipes?

Presumably for `ska_helpers` we only need to change the build recipes for namespace packages that use it?

**Comment by javierggt** [comment]
We can do whatever you prefer. I already changed all the recipes for namespace packages (https://github.com/sot/skare3/pull/977). I don't mind either way

**Comment by taldcroft** [comment]
FYI I haven't gotten to it yet but we agreed to leave the package `name` and repo name as `Ska.Sun` to minimize other disruptions. But I'm still fully open to ideas on this.

**Comment by taldcroft** [comment]
@javierggt @jeanconn - I updated the description to reflect testing using the correct install method and with the latest commit. I think this is ready for final review.

File: setup.py [hunk 1]
```diff
@@ -1,20 +1,31 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
+
 from setuptools import setup
+from testr.setup_helper import cmdclass
```
Comment (jeanconn) [inline setup.py:4]: I thought we were just going to cut this, but I don't have skin in the game.

Comment (taldcroft) [reply setup.py:4]: I think we said it might be a good thing in general, but that change is not needed for this PR so we can revisit.

**Review by jeanconn — APPROVED** [review]

**Pull request merged**